### PR TITLE
Sue spi wip

### PR DESCRIPTION
**Pull request by lyakh**
Early stage of SUE Creek SPI support. Note, new are only the last two patches.

### CODE REVIEW

File: src/drivers/intel/cavs/spi.c [hunk 1]
```diff
@@ -426,36 +447,66 @@ struct sspi *sspi_get(uint32_t type)
...
+
+		rmemcpy(config->src_buf, data, chunk_size);
+		dcache_writeback_region(config->src_buf, chunk_size);
+		sspi_trigger(spi, SSPI_TRIGGER_START, SPI_DIR_TX);
+		data += chunk_size;
+		size -= chunk_size;
+		gpio_write(GPIO14, 1);
+		delay(1);
+		gpio_write(GPIO14, 0);
+		/* 0.5s */
```
Comment (plbossart) [inline src/drivers/intel/cavs/spi.c]: if those GPIO toggles are for debug only use a helper function that is commented out if it's not needed?

File: src/drivers/intel/dw-dma.c [hunk 2]
```diff
@@ -1271,7 +1277,7 @@ static inline void dw_dma_interrupt_unregister(struct dma *dma, int channel)
 static int dw_dma_probe(struct dma *dma)
 {
 	struct dma_pdata *dw_pdata;
-	int i;
+	int i, ret;
```
Comment (plbossart) [inline src/drivers/intel/dw-dma.c]: one line per variable please.

File: src/drivers/intel/sue-gpio.c [hunk 3]
```diff
@@ -0,0 +1,121 @@
+/*
```
Comment (plbossart) [inline src/drivers/intel/sue-gpio.c]: why isn't in in intel/cavs?

File: src/include/sof/dma.h [hunk 4]
```diff
@@ -71,6 +71,7 @@
 #define DMA_DEV_HDA		BIT(1) /**< connectable to HD/A link */
 #define DMA_DEV_SSP		BIT(2) /**< connectable to SSP fifo */
 #define DMA_DEV_DMIC		BIT(3) /**< connectable to DMIC fifo */
+#define DMA_DEV_SSI		BIT(4) /**< connectable to SSI / SPI fifo */
```
Comment (plbossart) [inline src/include/sof/dma.h]: SSI is a Freescale/NXP interface, we'd want to have something less ambiguous here. This is really Intel-specific

File: src/platform/intel/cavs/dma.c [hunk 5]
```diff
@@ -160,6 +160,7 @@ struct dma dma[CAVS_PLATFORM_NUM_DMACS] = {
 	},
 	.ops		= &dw_dma_ops,
 },
+#if 0
```
Comment (plbossart) [inline src/platform/intel/cavs/dma.c]: remove code

File: src/platform/intel/cavs/memory.c [hunk 6]
```diff
@@ -105,7 +105,7 @@ struct mm memmap = {
 		.size = HEAP_RUNTIME_SIZE,
 		.info = {.free = HEAP_RUNTIME_SIZE,},
 		.caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_EXT |
-			SOF_MEM_CAPS_CACHE,
+			SOF_MEM_CAPS_CACHE | SOF_MEM_CAPS_DMA,
```
Comment (plbossart) [inline src/platform/intel/cavs/memory.c]: this feels wrong... Is SueCreek the only platform that needs this? Or are you using the wrong memory allocator?

File: src/platform/intel/cavs/platform.c [hunk 7]
```diff
@@ -367,8 +371,23 @@ int platform_init(struct sof *sof)
...
+	if (ret < 0)
+		return ret;
+
+	/* initialize the SPI-SLave module */
+	ret = spi_slave_init(spi, SOF_SPI_INTEL_SLAVE);
+	if (ret < 0)
+		return ret;
+
 	/* Initialize DMA for Trace*/
+#if !defined(CONFIG_SUECREEK)
```
Comment (plbossart) [inline src/platform/intel/cavs/platform.c]: it's more that we need a define to disable the trace, and that's set by default in SueCreek.
I'd love to disable the stupid trace, it's been problematic for a year now...

File: src/drivers/intel/dw-uart.c [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+					(uint8_t)(divisor & 0xFF));
+		uart_write(dev, SUE_UART_REG_BRDH,
+					(uint8_t)((divisor >> 8) & 0xFF));
+
+		/* restore to access baudrate divisor registers */
+		uart_write(dev, SUE_UART_REG_LCR, tmp);
+	}
+
+	/* 8Bit-data, 1-stop bit, no parity, clear DLAB */
+	uart_write(dev, SUE_UART_REG_LCR,
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/drivers/intel/dw-uart.c]: yep, should be platform data

File: src/drivers/intel/cavs/spi.c [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Zhigang Wu <zhigang.wu@intel.com>
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/drivers/intel/cavs/spi.c]: this is a generic driver (not for cavs only as is DW IP) and should be drivers/dw-spi.c

File: src/include/sof/spi.h [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+#define	SUE_SPI_REG_TXOICR		SUE_SPI_BASEADDRESS(0x38)
+#define	SUE_SPI_REG_RXOICR		SUE_SPI_BASEADDRESS(0x3C)
+#define	SUE_SPI_REG_RXUICR		SUE_SPI_BASEADDRESS(0x40)
+#define	SUE_SPI_REG_ICR			SUE_SPI_BASEADDRESS(0x48)
+#define	SUE_SPI_REG_DMACR		SUE_SPI_BASEADDRESS(0x4C)
+#define	SUE_SPI_REG_DMATDLR		SUE_SPI_BASEADDRESS(0x50)
+#define	SUE_SPI_REG_DMARDLR		SUE_SPI_BASEADDRESS(0x54)
+#define	SUE_SPI_REG_DR			SUE_SPI_BASEADDRESS(0x60)
+#define	SUE_SPI_REG_SPICTRLR0	SUE_SPI_BASEADDRESS(0xF4)
+
```
Comment (lgirdwood) [inline src/include/sof/spi.h]: This is generic DW IP, so we can remove the SUE_ and use platform data to set the base address.

File: src/platform/suecreek/include/platform/sue-gpio.h [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Wu Zhigang <zhigang.wu@intel.com>
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/platform/suecreek/include/platform/sue-gpio.h]: DW-IP - needs renames to dw-gpio.c, again with generic macros.

**Comment by lgirdwood** [comment]
@lyakh please reopen when you have new updates, I'll close for the moment.

**Pull request closed without merging**